File: codeforces/1800~1899/1891/c.cc
#include <bits/stdc++.h>

using namespace std;

using i64 = long long;

void solve() {
  int n = 0;
  cin >> n;
  vector<i64> a(n);
  for (int i = 0; i < n; ++i) cin >> a[i];

  sort(a.begin(), a.end());
  i64 x = 0;
  i64 ans = 0;
  int l = 0, r = n-1;
  while (l < r) {
    while (l < r && x < a[r]) {
      i64 require = a[r] - x;
      i64 op1 = min(a[l], require);
      a[l] -= op1;
      x += op1;
      ans += op1;
      if (a[l] == 0) l++;
    }
    if (x == a[r]) {
      // op2
      ans++;
      r--;
      x = 0;
    } else {
      break;
    }
  }
  if (l == r) {
    // 1 horde left
    i64 h = a[l];
    // cout << "x=" << x << ", h=" << h << endl;
    if (h == 1) {
      ans++;
    } else {
      i64 a = (h - x) / 2;
      ans += a;
      h -= (x + a + a);
      ans++;
      ans += h;
    }
  }

  cout << ans << endl;
}

int main(void) {
  ios::sync_with_stdio(false);
  cin.tie(0);

  int T = 0;
  cin >> T;
  for (int t = 0; t < T; ++t) {
    solve();
  }
  return 0;
}

File: codeforces/1800~1899/1891/c.rs
use std::io::{stdin, stdout, BufWriter, Write};
use std::str::FromStr;
use std::fmt::Debug;
use std::cmp::min;

fn read_int<T: FromStr>() -> T where <T as FromStr>::Err: Debug {
    let mut s = String::new();
    stdin().read_line(&mut s).unwrap();
    s.trim().parse::<T>().expect("invalid input")
}

fn read_vec<T: FromStr>() -> Vec<T> where <T as FromStr>::Err: Debug {
    let mut s = String::new();
    stdin().read_line(&mut s).unwrap();
    s.split(" ").map(|x| x.trim().parse().expect("invalid input")).collect()
}

fn solve() {
    let stdout = stdout();
    let mut writer = BufWriter::new(stdout.lock());
    let n = read_int::<i32>();
    let mut a = read_vec::<i64>();
    a.sort();

    let mut l = 0;
    let mut r = n-1;
    let mut x = 0;
    let mut ans = 0;
    while l < r {
        while l < r && x < a[r as usize] {
            let required = a[r as usize] - x;
            let op1 = min(required, a[l as usize]);
            x += op1;
            ans += op1;
            a[l as usize] -= op1;
            if a[l as usize] == 0 {
                l += 1;
            }
        }

        if x < a[r as usize] {
            break;
        }
        r -= 1;
        x = 0;
        ans += 1;
    }

    if l == r {
        let mut h = a[l as usize];
        if h == 1 {
            ans += 1;
        } else {
            let att = (h - x) / 2;
            ans += att;
            h -= x + att + att;
            ans += 1;
            ans += h;
        }
    }

    writeln!(writer, "{}", ans).ok();
}

fn main() {
    let n = read_int::<i32>();
    for _t in 0..n {
        solve();
    }
}
